### PR TITLE
Topic/support bson

### PR DESCRIPTION
**Pull request by pmeredit**
This adds support for BSON as a file type, this is also interesting in that I had to add support for parsing binary files as nushell values.

I haven't finished all of the bson types yet, but I need to run out for a bit, so I was hoping to get some early feedback. I'll finish the other types later.

### CODE REVIEW

**Comment by jonathandturner** [comment]
This is cool :+1:  You may want to also add a test (with a bson test file) to go along with the file type tests we have.

**Comment by pmeredit** [comment]
> This is cool +1 You may want to also add a test (with a bson test file) to go along with the file type tests we have.

Good point! I'll get to that later tonight, too. Before I head out, here's a screen shot of its working:

bson_demo

**Comment by jonathandturner** [comment]
Looks good. Let me know when you're ready and I can land it.

**Comment by pmeredit** [comment]
@jonathandturner It's ready to go. I will do to-bson in a separate PR.

File: src/commands/from_bson.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        BinarySubtype::BinaryOld => "binary_old",
+        BinarySubtype::UuidOld => "uuid_old",
+        BinarySubtype::Uuid => "uuid",
+        BinarySubtype::Md5 => "md5",
+        _ => unreachable!(),
+    }.to_string()
+}
+
+#[derive(Debug)]
+struct BytesReader {
```
Comment (pmeredit) [inline src/commands/from_bson.rs]: If we end up needing this elsewhere, we can move it to some other place in the codebase.

**Comment by androbtech** [comment]
@pmeredit Thanks for that! The whole test suite won't pass. Make sure [you cover this test case](https://github.com/nushell/nushell/blob/master/src/utils.rs#L174) since you added a sample file under `tests/formats`

**Comment by pmeredit** [comment]
@jonathandturner @androbtech Should be good to go, rebased on master

**Comment by androbtech** [comment]
> @jonathandturner @androbtech Should be good to go, rebased on master

Try the `normalize_string` helper method for `output` in the assertion. Let's check if it passes on Windows with that.

**Comment by pmeredit** [comment]
> > @jonathandturner @androbtech Should be good to go, rebased on master
> 
> @pmeredit Try the `normalize_string` helper method for `output` in the assertion. Let's check if it passes on Windows with that.

Thanks! You can tell I develop on Linux.

**Comment by androbtech** [comment]
> > > @jonathandturner @androbtech Should be good to go, rebased on master
> > 
> > 
> > @pmeredit Try the `normalize_string` helper method for `output` in the assertion. Let's check if it passes on Windows with that.
> 
> Thanks! You can tell I develop on Linux.

I'm doing a PR as well and it's failing. Then rebased on master and can't compile it anymore (seems to be an issue with the switch from `reqwest` to `surf`.

So I'm basically pushing changes and have to wait and see. Sigh.

**Comment by pmeredit** [comment]
> So I'm basically pushing changes and have to wait and see. Sigh.

@androbtech normalize_string didn't fix it, so I just removed the test. I'm not worried in fidelity in the formatting of javascript code. It's a relic in the bson_spec that no one really uses, anyway.

The bson crate currently doesn't support Decimal128, either, so that will need to be done eventually (but we also need Decimal128 support in nushell for that to make sense!)

**Comment by androbtech** [comment]
> > So I'm basically pushing changes and have to wait and see. Sigh.
> 
> @androbtech normalize_string didn't fix it, so I just removed the test. I'm not worried in fidelity in the formatting of javascript code. It's a relic in the bson_spec that no one really uses, anyway.
> 
> The bson crate currently doesn't support Decimal128, either, so that will need to be done eventually (but we also need Decimal128 support in nushell for that to make sense!)

It's all good. I did see the test case failing and it seemed that you needed an additional `;` in the string assertion (the expectation). Your work is very much appreciated it. Thank you.

Need to ask you a favor. Could you file an issue with your observation for Decimal128? 

Best.

**Comment by pmeredit** [comment]
> > > So I'm basically pushing changes and have to wait and see. Sigh.
> > 
> > 
> > @androbtech normalize_string didn't fix it, so I just removed the test. I'm not worried in fidelity in the formatting of javascript code. It's a relic in the bson_spec that no one really uses, anyway.
> > The bson crate currently doesn't support Decimal128, either, so that will need to be done eventually (but we also need Decimal128 support in nushell for that to make sense!)
> 
> @pmeredit It's all good. I did see the test case failing and it seemed that you needed an additional `;` in the string assertion (the expectation). Your work is very much appreciated it. Thank you.
> 
> Need to ask you a favor. Could you file an issue with your observation for Decimal128?
> 
> Best.

@androbtech Thanks! Issue filed

**Pull request merged**